File: PAT_Basic_Level_by_c/1114 全素日/1114 全素日.c
/*
 * 题目：1114 全素日
 * 实现原理：
 *   判断日期字符串是否为全素日。全素日定义：日期的每一位前缀都是素数。
 *   例如：23332，依次检查2、23、233、2333、23332是否都是素数。
 *   算法步骤：
 *   1. is_prime函数：判断一个数是否为素数
 *   2. 读取日期字符串
 *   3. 依次检查从最长到最短的前缀是否为素数
 *   4. 如果所有前缀都是素数，输出"All Prime!"
 */
#include <stdio.h>
#include <string.h>
#include <math.h>

int is_prime(int x) {       // 判断 x 是否为素数
    if (x < 2) return 0;
    for (int i = 2; i * i <= x; i++) {
        if (x % i == 0) return 0;
    }
    return 1;
}

int main() {
    char s[20];
    scanf("%s", s);         // 读入日期字符串
    int len = strlen(s);
    int all_prime = 1;      // 标记所有前缀是否都为素数
    while (len > 0) {       // 依次检查从最长到最短的前缀
        int num = 0;
        for (int i = 0; i < len; i++) {     // 将当前前缀转为整数
            num = num * 10 + (s[i] - '0');
        }
        printf("%s %s\n", s, is_prime(num) ? "Yes" : "No");
        if (!is_prime(num)) all_prime = 0;
        for (int i = 0; i < len - 1; i++) { // 去掉首字符，得到下一个更短的前缀
            s[i] = s[i + 1];
        }
        s[len - 1] = '\0';
        len--;
    }
    if (all_prime) printf("All Prime!\n");  // 所有前缀均为素数
    return 0;
}

File: PAT_Basic_Level_by_ruby/1114 全素日/1114 全素日.rb
# 1114 全素日
#
# 题目描述：
#   ![wbfg.JPG](~/f4c6f880-b7ed-4893-b10f-045937586870.JPG)
#   以上图片来自新浪微博，展示了一个非常酷的“全素日”：2019年5月23日。即不仅`20190523`本身是个素数，它的任何以末尾数字`3`结尾的子串都是素数。
#   本题就请你写个程序判断一个给定日期是否是“全素日”。
#
# 实现原理：
#   判断一个日期是否所有部分都是素数；依次检查年、月、日及其组合是否为素数；素数判断和日期验证
#

if __FILE__ == $PROGRAM_NAME
  def prime?(number)
    return false if number < 2
    return true if number == 2
    return false if number.even?

    (3..Math.sqrt(number).to_i).step(2).none? { |d| (number % d).zero? }
  end

  date = gets.to_s.chomp
  all_prime = true
  date.length.downto(1) do |length|
    part = date[date.length - length, length]
    is_prime = prime?(part.to_i)
    all_prime = false unless is_prime
    puts "#{part} #{is_prime ? 'Yes' : 'No'}"
  end
  puts 'All Prime!' if all_prime
end
